src/msg_recorder/record_file.cc

```diff
@@ -0,0 +1,119 @@
+#include "cris/core/msg_recorder/record_file.h"
+
+#include "cris/core/utils/defs.h"
+#include "cris/core/utils/logging.h"
+#include "cris/core/utils/time.h"
+
+#include <algorithm>
+#include <filesystem>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace cris::core {
+
+RecordFileKey RecordFileKey::Make() {
+    static std::atomic<unsigned long long> global_counter = 0;
+
+    return {
+        .timestamp_ = GetUnixTimestampNsec(),
+        .count_     = global_counter.fetch_add(1),
+    };
+}
+
+RecordFileKey RecordFileKey::FromSlice(const leveldb::Slice& slice) {
+    RecordFileKey key;
+    std::memcpy(&key, slice.data(), std::min(sizeof(key), slice.size()));
+    return key;
+}
+
+int RecordFileKey::compare(const RecordFileKey& lhs, const RecordFileKey& rhs) {
+    if (lhs.timestamp_ < rhs.timestamp_) {
+        return -1;
+    } else if (lhs.timestamp_ > rhs.timestamp_) {
+        return 1;
+    } else if (lhs.count_ < rhs.count_) {
+        return -1;
+    } else if (lhs.count_ > rhs.count_) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+int RecordFileKey::LevelDBComparator::Compare(const leveldb::Slice& lhs, const leveldb::Slice& rhs) const {
+    return RecordFileKey::compare(RecordFileKey::FromSlice(lhs), RecordFileKey::FromSlice(rhs));
+}
+
+const char* RecordFileKey::LevelDBComparator::Name() const {
+    static const auto name = GetTypeName<std::remove_cvref_t<decltype(*this)>>();
+    return name.c_str();
+}
+
+RecordFileIterator::RecordFileIterator(leveldb::Iterator* db_itr) : db_itr_(db_itr) {
+}
+
+bool RecordFileIterator::Valid() const {
+    return db_itr_->Valid();
+}
+
+RecordFileKey RecordFileIterator::GetKey() const {
+    return RecordFileKey::FromSlice(db_itr_->key());
+}
+
+std::pair<RecordFileKey, std::string> RecordFileIterator::Get() const {
+    return std::make_pair(GetKey(), db_itr_->value().ToString());
+}
+
+void RecordFileIterator::Next() {
+    db_itr_->Next();
+}
+
+RecordFile::RecordFile(std::string file_path) : file_path_(std::move(file_path)) {
```

Comment:
Saw you copy and immediately move in many places, why? Why not const&?

Reply:
Because there is no copy in this way, only move ctor is called.

Reply:
Only if you called with rv initially.
If you call with lv, there'll be a copy on arg.
If you want to fully eliminate the copy, you need perfect forwarding with template.

Reply:
> Only if you called with rv initially.
> If you call with lv, there'll be a copy on arg.

If calling with lv, there will be copy whatever you do, and we only do a few more moves here.

> If you want to fully eliminate the copy, you need perfect forwarding with template.

PF is not a silver bullet, and it does not make much difference in this case.